packages/design-system/src/components/Button/Button.tsx

```diff
@@ -1,74 +1,45 @@
-import React from 'react';
-import { IconName } from '@talend/icons';
-import { ClickableProps } from 'reakit';
-
-import { Icon } from '../Icon/Icon';
-import Loading from '../Loading';
-import Tooltip, { TooltipPlacement } from '../Tooltip';
-
-import * as S from './Button.style';
-
-type BaseProps = {
-	/** The icon of the button */
-	icon?: IconName | React.ReactElement;
-	/** If the button is small or not */
-	small?: boolean;
-	/** If the button is loading or not */
-	loading?: boolean;
-	/** If the button should not display text */
-	hideText?: boolean;
-	/** All buttons must have contents */
-	children: React.ReactNode | React.ReactNodeArray;
-	/** Use these if the button should be an anchor or router link */
-	as?: React.ElementType;
-	/** When hideText is set, a tooltip appears. Use this if you want to customize its position */
-	tooltipPlacement?: TooltipPlacement;
-	href?: string;
-	target?: string;
-};
-
-export type ButtonProps = ClickableProps & BaseProps;
-
-const Button = React.forwardRef(
-	(
-		{ className, icon, small, hideText, loading, children, tooltipPlacement, ...rest }: ButtonProps,
-		ref: React.Ref<any>,
-	) => {
-		const button = (
-			<S.Button
-				ref={ref}
-				{...rest}
-				className={`
-				btn ${className || ''} ${icon ? 'btn--has-icon' : ''} ${hideText ? '' : 'btn--has-text'} ${
-					small ? 'btn--small' : ''
-				} ${loading ? 'btn--loading' : ''}
-			`}
-				aria-busy={!!loading}
-			>
-				{loading && <Loading className="btn__loading btn__icon" name={icon} aria-hidden />}
-				{!loading &&
-					icon &&
-					(typeof icon === 'string' ? (
-						<Icon className="btn__icon" name={icon} />
-					) : (
-						React.cloneElement(icon, {
-							className: `${icon.props?.className} btn__icon`,
-						})
-					))}
-				<span className={`btn__text ${hideText ? 'btn__text--hidden' : ''}`}>{children}</span>
-			</S.Button>
-		);
-
-		return hideText ? (
-			<Tooltip title={children} placement={tooltipPlacement}>
-				{button}
-			</Tooltip>
-		) : (
-			button
-		);
-	},
-);
-
-Button.displayName = 'Button';
+import React, { forwardRef, Ref } from 'react';
+
+import ButtonPrimary, { ButtonPrimaryPropsType } from './variations/ButtonPrimary';
+import ButtonSecondary, { ButtonSecondaryPropsType } from './variations/ButtonSecondary';
+import ButtonTertiary, { ButtonTertiaryPropsType } from './variations/ButtonTertiary';
+import ButtonDestructive, { ButtonDestructivePropsType } from './variations/ButtonDestructive';
+import { availableVariants } from './Primitive/ButtonPrimitive';
+import Clickable from '../Clickable';
+
+type Primary = { variant: typeof availableVariants.primary } & ButtonPrimaryPropsType;
```

Comment:
Typing availableVariants as an enum would be better and remove the need of typeof here

Reply:
Strong disagree. Enums in TS are overkill and create poor dev experience. For TS to be happy, when using your Button, you will need to feed it the Enum. I don't see the point. Typing isn't stronger, autocomplete is not better etc...

Reply:
use string unions then, don't create an object if it's only used with a typeof...

Reply:
Fine by me

Reply:
or enum + template if you want to expose an object...

Reply:
I don't understand why enums would ever be better honestly, in this prop-based context.

Reply:
Fixed with string and generic types